src/aspire/ctf/ctf_estimator.py

```diff
@@ -0,0 +1,757 @@
+
+        moment_mat = np.zeros((2, 2))
+        moment_mat[0, 0] = moment_20
+        moment_mat[1, 1] = moment_02
+        moment_mat[0, 1] = moment_11
+        moment_mat[1, 0] = moment_11
+
+        moment_evals, moment_evecs = npla.eigh(moment_mat)
+        ratio = np.divide(moment_evals[0], moment_evals[1])
+
+        return ratio
+
+    def ctf_gd(
+        self,
+        signal,
+        df1,
+        df2,
+        angle_ast,
+        r,
+        theta,
+        pixel_size,
+        g_min,
+        g_max,
+        amplitude_contrast,
+        lmbd,
+        cs,
+    ):
+        """
+        Runs gradient ascent to optimize defocus parameters
+
+        :param signal: Estimated power spectrum
+        :param df1: Defocus value in the direction perpendicular to df2.
+        :param df2: Defocus value in the direction perpendicular to df1.
+        :param angle_ast: Angle between df1 and the x-axis
+        :param r: Magnitude of spatial frequencies.
+        :param theta: Phase of spatial frequencies.
+        :param pixel_size: Pixel size in angstrom.
+        :param g_min: Inverse of minimun resolution for PSD.
+        :param g_max: Inverse of maximum resolution for PSD.
+        :param amplitude_contrast: Amplitude contrast.
+        :param lmbd: Electron wavelength.
+        :param cs: Spherical aberration in mm.
+        :return: Optimal defocus parameters
+        """
+
+        angle_ast = angle_ast / 180 * np.pi
+
+        # step size
+        alpha1 = np.power(10, 5)
+        alpha2 = np.power(10, 4)
+
+        # initialization
+        x = df1 + df2
+        y = (df1 - df2) * np.cos(2 * angle_ast)
+        z = (df1 - df2) * np.sin(2 * angle_ast)
+
+        a = np.pi * lmbd * np.power(r, 2) / 2
+        b = np.pi * np.power(lmbd, 3) * cs * np.power(10, 6) * np.power(
+            r, 4
+        ) / 2 - amplitude_contrast * np.ones(r.shape)
+
+        signal = signal.asnumpy()[0].T
+        N = signal.shape[1]
+        center = N // 2
+
+        rad_sq_min = N * pixel_size / g_min
+        rad_sq_max = N * pixel_size / g_max
+
+        max_val = r[center, np.int(center - 1 + np.floor(rad_sq_max))]
+        min_val = r[center, np.int(center - 1 + np.ceil(rad_sq_min))]
+        valid_region = np.argwhere(
+            np.logical_and(np.where(r <= max_val, 1, 0), np.where(r > min_val, 1, 0))
+        )
+
+        a = a[valid_region[:, 0], valid_region[:, 1]]
+        b = b[valid_region[:, 0], valid_region[:, 1]]
+        signal = signal[..., valid_region[:, 0], valid_region[:, 1]]
+        r = r[valid_region[:, 0], valid_region[:, 1]]
+        theta = theta[valid_region[:, 0], valid_region[:, 1]]
+
+        sum_A = np.sum(np.power(signal, 2))
+
+        dx = 1
+        dy = 1
+        dz = 1
+
+        stop_cond = 10 ^ -20
+        iter_no = 1
+
+        # for iter_no in range(0, 399):
+        while np.maximum(np.maximum(dx, dy), dz) > stop_cond:
+            inner_cosine = np.multiply(y, np.cos(2 * theta)) + np.multiply(
```

Comment:
These formulas all need to be documented somewhere at some point… To start with, it'll be easier to get an idea of what's going on if all the function calls are replaced with operators, though.

Reply:
I replaced the functions with the operators for `gd`. It was a big change, 3ee595

Reply:
Much better.